### PR TITLE
Allow dynamic linking to system libuv

### PR DESCRIPTION
**Pull request by mtorromeo**
Added build_ext option --use-system-libuv to switch between dynamic and static linking.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -17,6 +17,16 @@
 
 
 class libuv_build_ext(build_ext):
+    build_ext.user_options.extend([
+        ("use-system-libuv", None, "Use the system provided libuv, instead of the bundled one")
```
Comment (1st1) [inline setup.py]: This line is too long.

Comment (mtorromeo) [inline setup.py]: Fixed

File: setup.py [hunk 2]
```diff
@@ -35,14 +45,17 @@ def build_libuv(self):
...
-        libuv_lib = os.path.join(LIBUV_DIR, '.libs', 'libuv.a')
-        if not os.path.exists(libuv_lib):
-            self.build_libuv()
-        if not os.path.exists(libuv_lib):
-            raise RuntimeError('failed to build libuv')
-
-        self.extensions[-1].extra_objects.extend([libuv_lib])
-        self.compiler.add_include_dir(os.path.join(LIBUV_DIR, 'include'))
+        if self.use_system_libuv:
+            self.compiler.add_library('uv')
```
Comment (1st1) [inline setup.py]: Can you also add this snippet of code here, to support Mac OS X + macports:

```python
            if sys.platform == 'darwin' and \
                    os.path.exists('/opt/local/include'):
                # Support macports on Mac OS X.
                self.compiler.add_include_dir('/opt/local/include')
```

**Comment by 1st1** [comment]
Thanks for the PR! BTW, uvloop requires libuv @v1.9.0.  It won't compile with an earlier version.  Is there any way to check the version of libuv in `setup.py`, or we should just let the compiler to fail?

File: setup.py (hunk 2 above)
Comment (mtorromeo) [inline setup.py]: Snippet added but I cannot test it since I don't use Mac OS X.

Comment (1st1) [inline setup.py]: I've tested it ;)

**Comment by mtorromeo** [comment]
I'm not sure how feasible it would be to test libuv version but I'd rather keep things simple.

The shared object version would be a way to check for ABI compatibility but version 1.8.0 installed `libuv.so.1.0.0` just as 1.9.0 does.

Having the compiler just fail doesn't seem too bad to me as long as it is documented somewhere since `--use-system-libuv` is explicit and you should know what you are doing. Distribution packagers know how to check these things.

**Comment by 1st1** [comment]
Alright, merged in d65265689baf9b7a51952e980ec870df4a3a660b. Thanks a lot!

**Comment by 1st1** [comment]
FYI `v0.4.16` has this PR in.

**Pull request closed without merging**